### PR TITLE
Change DMReportGenerator to accommodate EdgeAgent Ping test

### PR DESCRIPTION
**Pull request by dylanbronson**
We are adding a test for EdgeAgent direct method "ping" method to Connectivity tests.

We won't have a ReceiverSource for DirectMethod in that case. So this PR makes ReceiverSource and ReceiverTestResults Optional. 

It also changes the implementation of the DMReportGeneration to accomodate when these Optional values are not filled in.

### CODE REVIEW

File: Microsoft.Azure.Devices.Edge.sln [hunk 1]
```diff
@@ -15,6 +15,7 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "hub", "hub", "{5037265C-7AB
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "util", "util", "{F7D7DD5B-42E1-4E62-8951-514A8711F96D}"
 	ProjectSection(SolutionItems) = preProject
+		test\modules\TestResultCoordinator\Reports\DirectMethod\DirectMethodReportGenerator.cs = test\modules\TestResultCoordinator\Reports\DirectMethod\DirectMethodReportGenerator.cs
```
Comment (philipktlin) [inline Microsoft.Azure.Devices.Edge.sln]: why has this change?

File: Microsoft.Azure.Devices.Edge.sln [hunk 2]
```diff
@@ -213,7 +214,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "TwinTester", "test\modules\
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "DeploymentTester", "test\modules\DeploymentTester\DeploymentTester.csproj", "{1C1C8203-CD73-4CC1-9112-7315F9DE1AB6}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Microsoft.Azure.Devices.Edge.Agent.Integration.Test", "edge-agent\test\Microsoft.Azure.Devices.Edge.Agent.Integration.Test\Microsoft.Azure.Devices.Edge.Agent.Integration.Test.csproj", "{B3C68519-B309-42FF-B4DF-9E77B2D4BD50}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Azure.Devices.Edge.Agent.Integration.Test", "edge-agent\test\Microsoft.Azure.Devices.Edge.Agent.Integration.Test\Microsoft.Azure.Devices.Edge.Agent.Integration.Test.csproj", "{B3C68519-B309-42FF-B4DF-9E77B2D4BD50}"
```
Comment (philipktlin) [inline Microsoft.Azure.Devices.Edge.sln]: why this change?

File: test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportMetadataJsonConverterTest.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+namespace Modules.Test.TestResultCoordinator.Reports.DirectMethod
+{
+    using System;
+    using global::TestResultCoordinator.Reports;
+    using global::TestResultCoordinator.Reports.DirectMethod;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil;
+    using Newtonsoft.Json;
+    using Xunit;
+
+    public class DirectMethodReportMetadataJsonConverterTest
```
Comment (philipktlin) [inline test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportMetadataJsonConverterTest.cs]: is this still valid? I can't find any DirectMethodReportMetadataJsonConverter.cs file in this PR.

File: test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportMetadataJsonConverterTest.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        const string SerializedTestJsonWithReceiverSource = "{ \"TestReportType\": \"DirectMethodReport\", \"SenderSource\": \"directMethodSender1.send\"," +
+            " \"ReceiverSource\": \"directMethodReceiver1.receive\", \"TolerancePeriod\": \"00:00:00.005\" }";
+        const string SerializedTestJsonWithoutReceiverSource = "{ \"TestReportType\": \"DirectMethodReport\", \"SenderSource\": \"directMethodSender1.send\", " +
+            "\"TolerancePeriod\": \"00:00:00.005\" }";
+
+        [Fact]
+        public void DirectMethodJsonConvertWitReceiverSourceTest()
```
Comment (philipktlin) [inline test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportMetadataJsonConverterTest.cs]: we have similar tests in TestReportUtilTest class.  Move these to there.

File: test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/ReportDataForOneSource.cs [hunk 5]
```diff
@@ -0,0 +1,138 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Modules.Test.TestResultCoordinator.Reports.DirectMethod
```
Comment (philipktlin) [inline test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/ReportDataForOneSource.cs]: This is used for test only, right? you should move it to Modules.Test project.

Comment (philipktlin) [reply test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/ReportDataForOneSource.cs]: Rename it to more specific, e.g. DirectMethodReportDataWithSenderSouceOnly and DirectMethodReportDataWithSenderAndReceiverSources.

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReport.cs [hunk 6]
```diff
@@ -64,7 +64,15 @@ class DirectMethodReport : TestResultReportBase
         // MismatchFailure is when the there is a result in ActualStore but no result in ExpectedStore. This should never happen.
         public ulong MismatchFailure { get; }
 
-        public override string Title => $"DirectMethod Report ({this.ResultType}) for [{this.ExpectedSource}] and [{this.ActualSource}]";
+        public override string Title
+        {
+            get
+            {
+                return this.ReceiverSource.Match<string>(
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReport.cs]: if you want to simplify, you can just call this.ReceiverSource.ToString().

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportMetadata.cs [hunk 7]
```diff
@@ -2,26 +2,45 @@
...
-        public string ActualSource { get; }
+        public Option<string> ReceiverSource { get; }
 
-        public string[] ResultSources => new string[] { this.ExpectedSource, this.ActualSource };
+        public string[] ResultSources
+        {
+            get
+            {
+                List<string> resultSources = new List<string>();
+                resultSources.Add(this.SenderSource);
```
Comment (yophilav) [inline test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportMetadata.cs]: For everytime `ResultSources` is accessed, add a `SenderSource` to the `resultSources`? I'm confused.

File: test/modules/Modules.Test/TestResultCoordinator/TestReportUtilTest.cs [hunk 8]
```diff
@@ -209,8 +209,9 @@ public void ParseReportMetadataList_ParseDirectMethodTestReport()
             Assert.NotNull(reportMetadata);
             Assert.Equal(TestOperationResultType.DirectMethod, reportMetadata.TestOperationResultType);
             Assert.Equal(TestReportType.DirectMethodReport, reportMetadata.TestReportType);
-            Assert.Equal("directMethodSender1.send", reportMetadata.ExpectedSource);
-            Assert.Equal("directMethodReceiver1.receive", reportMetadata.ActualSource);
+            Assert.Equal("directMethodSender1.send", reportMetadata.SenderSource);
+            Assert.True(reportMetadata.ReceiverSource.HasValue);
+            reportMetadata.ReceiverSource.ForEach(x => Assert.Equal("directMethodReceiver1.receive", x));
```
Comment (yophilav) [inline test/modules/Modules.Test/TestResultCoordinator/TestReportUtilTest.cs]: Adding a comment b/c I didn't mean to approve it.

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReport.cs [hunk 9]
```diff
@@ -11,8 +11,8 @@ class DirectMethodReport : TestResultReportBase
     {
         public DirectMethodReport(
             string trackingId,
-            string expectedSource,
-            string actualSource,
+            string senderSource,
+            Option<string> receiverSource,
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReport.cs]: If serialize this report, the json for receiverSource is an object containing Value and HasValue properties.  Or should we use string and allows empty string instead.  You decide.

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs [hunk 10]
```diff
@@ -21,31 +21,41 @@ sealed class DirectMethodReportGenerator : ITestResultReportGenerator
...
-            ITestResultCollection<TestOperationResult> actualTestResults,
+            string senderSource,
+            ITestResultCollection<TestOperationResult> senderTestResults,
+            Option<string> receiverSource,
+            Option<ITestResultCollection<TestOperationResult>> receiverTestResults,
             string resultType,
             ITestResultComparer<TestOperationResult> testResultComparer,
             NetworkStatusTimeline networkStatusTimeline)
         {
+            if (receiverSource.HasValue && !receiverTestResults.HasValue)
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs]: just to combine; and just throw argument exception with same message; e.g. "please provide both receiver source and receiver test results or neither".

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs [hunk 11]
```diff
@@ -66,89 +76,134 @@ public async Task<ITestResultReport> CreateReportAsync()
...
-                        this.NetworkStatusTimeline.GetNetworkControllerStatusAndWithinToleranceAt(this.ExpectedTestResults.Current.CreatedAt);
-                if (!NetworkControllerStatus.Enabled.Equals(networkControllerStatus) &&
-                    !NetworkControllerStatus.Disabled.Equals(networkControllerStatus))
+                    this.NetworkStatusTimeline.GetNetworkControllerStatusAndWithinToleranceAt(this.SenderTestResults.Current.CreatedAt);
+                this.ValidateNetworkControllerStatus(networkControllerStatus);
+                DirectMethodTestResult dmSenderTestResult = JsonConvert.DeserializeObject<DirectMethodTestResult>(this.SenderTestResults.Current.Result);
+                if (hasReceiverResult)
                 {
-                    throw new InvalidOperationException($"Unexpected Result. NetworkControllerStatus was {networkControllerStatus}");
+                    string receiverSource = this.ReceiverSource.Expect<ArgumentException>(
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs]: just call this.ReceiverSource.OrDefault() since you already validate in constructor.

Comment (philipktlin) [reply test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs]: same for this.ReceiverTestResults

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodTestOperationResultComparer.cs [hunk 12]
```diff
@@ -33,7 +33,8 @@ public bool Matches(TestOperationResult value1, TestOperationResult value2)
             DirectMethodTestResult dmtr1 = JsonConvert.DeserializeObject<DirectMethodTestResult>(value1.Result);
             DirectMethodTestResult dmtr2 = JsonConvert.DeserializeObject<DirectMethodTestResult>(value2.Result);
 
-            return string.Equals(dmtr1.SequenceNumber, dmtr2.SequenceNumber, StringComparison.OrdinalIgnoreCase);
+            return string.Equals(dmtr1.SequenceNumber, dmtr2.SequenceNumber, StringComparison.OrdinalIgnoreCase)
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodTestOperationResultComparer.cs]: This is not an operation result comparer; I think it should rename as sth for matching message identity, e.g. DirectMethodTestOperationResultIdentityComparer.

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs [hunk 13]
```diff
@@ -66,89 +76,134 @@ public async Task<ITestResultReport> CreateReportAsync()
...
+                        }
+                        else
+                        {
+                            throw new InvalidDataException($"Sequence numbers should not match if the testResults didn't match. SenderTestResult: " +
+                                $"{dmSenderTestResult.GetFormattedResult()}. ReceiverTestResult: {dmReceiverTestResult.GetFormattedResult()}");
+                        }
+                    }
+                    else
+                    {
+                        hasReceiverResult = await receiverTestResults.MoveNextAsync();
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs]: you should only move receiver result pointer when matched, right?

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs [hunk 14]
```diff
@@ -66,89 +76,134 @@ public async Task<ITestResultReport> CreateReportAsync()
...
+                DirectMethodTestResult dmSenderTestResult = JsonConvert.DeserializeObject<DirectMethodTestResult>(this.SenderTestResults.Current.Result);
+                if (hasReceiverResult)
                 {
-                    throw new InvalidOperationException($"Unexpected Result. NetworkControllerStatus was {networkControllerStatus}");
+                    string receiverSource = this.ReceiverSource.Expect<ArgumentException>(
+                        () => throw new ArgumentException("There must be a receiver source if there are receiver test results"));
+                    ITestResultCollection<TestOperationResult> receiverTestResults = this.ReceiverTestResults.Expect<ArgumentException>(
+                        () => throw new ArgumentException("Previous check for receiver source passed"));
+                    this.ValidateDataSource(receiverTestResults.Current, receiverSource);
+                    if (!this.TestResultComparer.Matches(receiverTestResults.Current, this.SenderTestResults.Current))
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs]: you need to think about it may be better to deserialize directly here instead of use comparer.

Comment (philipktlin) [reply test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs]: Let's do pairing together tomorrow.

File: e2e_deployment_files/connectivity_deployment.template.json [hunk 15]
```diff
@@ -602,69 +602,74 @@
...
-              "TestReportType": "DirectMethodReport",
-              "ExpectedSource": "directMethodSender2.send",
-              "ActualSource": "directMethodReceiver2.receive",
-              "TolerancePeriod": "00:00:00.005"
+            "TestReportType": "DirectMethodReport",
+            "SenderSource": "directMethodSender2.send",
+            "ReceiverSource": "directMethodReceiver2.receive",
+            "TolerancePeriod": "00:00:00.005"
+          },
+          "reportMetadata12": {
```
Comment (philipktlin) [inline e2e_deployment_files/connectivity_deployment.template.json]: thanks for fixing spacing.

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportMetadata.cs [hunk 16]
```diff
@@ -2,26 +2,45 @@
...
     {
-        public DirectMethodReportMetadata(string expectedSource, string actualSource, TestReportType testReportType, TimeSpan tolerancePeriod)
+        public DirectMethodReportMetadata(string senderSource, TestReportType testReportType, TimeSpan tolerancePeriod, string receiverSource = "")
         {
-            this.ExpectedSource = expectedSource;
-            this.ActualSource = actualSource;
+            this.SenderSource = senderSource;
             this.TestReportType = testReportType;
             this.TolerancePeriod = tolerancePeriod;
+            if (receiverSource is null || string.Empty.Equals(receiverSource))
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportMetadata.cs]:
```suggestion
            if (string.IsNullOrEmpty(receiverSource))
```

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportMetadata.cs [hunk 17]
```diff
@@ -2,26 +2,45 @@
...
 
-        public string ActualSource { get; }
+        public Option<string> ReceiverSource { get; }
 
-        public string[] ResultSources => new string[] { this.ExpectedSource, this.ActualSource };
+        public string[] ResultSources
+        {
+            get
+            {
+                List<string> resultSources = new List<string>();
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportMetadata.cs]:
```suggestion
                var resultSources = new List<string>();
```

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportMetadata.cs [hunk 18]
```diff
@@ -2,26 +2,45 @@
...
+                resultSources.Add(this.SenderSource);
+                this.ReceiverSource.ForEach(x => resultSources.Add(x));
+                return resultSources.ToArray();
+            }
+        }
 
         public TestReportType TestReportType { get; }
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportMetadata.cs]: you can hard-code test report type here and don't need to get from constructor.

File: test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/ReportDataForOneSource.cs (hunk 5 above)
Comment (dylanbronson) [reply test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/ReportDataForOneSource.cs]: This is already in Modules.Test project. What do you mean?

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReport.cs (hunk 6 above)
Comment (dylanbronson) [reply test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReport.cs]: I think this prints out more nicely and is not so complicated

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReport.cs (hunk 9 above)
Comment (dylanbronson) [reply test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReport.cs]: I like option because it explicitly lets us know that ReceiverSource is optional.

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportMetadata.cs (hunk 7 above)
Comment (dylanbronson) [reply test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportMetadata.cs]: Changed around to make it more readable

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs [hunk 19]
```diff
@@ -66,89 +76,134 @@ public async Task<ITestResultReport> CreateReportAsync()
...
+                                $" {receiverTestResults.Current.Source}," +
+                                $" {receiverTestResults.Current.Type}, " +
+                                $"{receiverTestResults.Current.Result} at " +
+                                $"{receiverTestResults.Current.CreatedAt}");
+                            hasReceiverResult = await receiverTestResults.MoveNextAsync();
+                            continue;
+                        }
+                        else if (int.Parse(dmSenderTestResult.SequenceNumber) < int.Parse(dmReceiverTestResult.SequenceNumber))
+                        {
+                            if (HttpStatusCode.OK.Equals((HttpStatusCode)int.Parse(dmSenderTestResult.Result)) &&
```
[the remark this answers is not in the document]
Comment (dylanbronson) [reply test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs]: Good idea - now using HttpStatusCode as DirectMethodTestResult's field

File: test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportMetadataJsonConverterTest.cs (hunk 3 above)
Comment (dylanbronson) [reply test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportMetadataJsonConverterTest.cs]: Deleted, moved tests to TestReportUtilTest

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs (hunk 14 above)
Comment (dylanbronson) [reply test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs]: Serializing directly. Got rid of comparer

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodTestOperationResultComparer.cs (hunk 12 above)
Comment (dylanbronson) [reply test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodTestOperationResultComparer.cs]: Removed whole class

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReport.cs (hunk 6 above)
Comment (dylanbronson) [reply test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReport.cs]: Nevermind - changed it a bit

File: Microsoft.Azure.Devices.Edge.sln (hunk 2 above)
Comment (dylanbronson) [reply Microsoft.Azure.Devices.Edge.sln]: changed back

File: test/modules/ModuleLib/TestResults/DirectMethodTestResult.cs [hunk 20]
```diff
@@ -13,13 +14,13 @@ public class DirectMethodTestResult : TestResultBase
...
             string sequenceNumber,
-            string result)
+            HttpStatusCode result)
             : base(source, TestOperationResultType.DirectMethod, createdAt)
         {
             this.TrackingId = trackingId ?? string.Empty;
             this.BatchId = Preconditions.CheckNotNull(batchId, nameof(batchId)).ToString();
             this.SequenceNumber = Preconditions.CheckNonWhiteSpace(sequenceNumber, nameof(sequenceNumber));
-            this.Result = Preconditions.CheckNonWhiteSpace(result, nameof(result));
+            this.Result = Preconditions.CheckNotNull(result, nameof(result));
```
Comment (philipktlin) [inline test/modules/ModuleLib/TestResults/DirectMethodTestResult.cs]:
```suggestion
            this.Result = result;
```

File: test/modules/ModuleLib/TestResults/DirectMethodTestResult.cs [hunk 21]
```diff
@@ -28,7 +29,7 @@ public class DirectMethodTestResult : TestResultBase
 
         public string SequenceNumber { get; set; }
 
-        public string Result { get; set; }
+        public HttpStatusCode Result { get; set; }
```
Comment (philipktlin) [inline test/modules/ModuleLib/TestResults/DirectMethodTestResult.cs]: nit: rename to "StatusCode"/"HttpStatusCode"

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReport.cs [hunk 22]
```diff
@@ -64,7 +64,8 @@ class DirectMethodReport : TestResultReportBase
         // MismatchFailure is when the there is a result in ActualStore but no result in ExpectedStore. This should never happen.
         public ulong MismatchFailure { get; }
 
-        public override string Title => $"DirectMethod Report for [{this.ExpectedSource}] and [{this.ActualSource}] ({this.ResultType})";
+        public override string Title => this.ReceiverSource.HasValue ?
+            $"DirectMethod Report ({this.ResultType}) for [{this.SenderSource}] and [{this.ReceiverSource.OrDefault()}]" : $"DirectMethod Report ({this.ResultType}) for [{this.SenderSource}]";
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReport.cs]:
```suggestion
            $"DirectMethod Report ({this.ResultType}) for [{this.SenderSource}] and [{this.ReceiverSource}]" : $"DirectMethod Report ({this.ResultType}) for [{this.SenderSource}]";
```

Comment (philipktlin) [inline test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReport.cs]: I already changed other reports to move result type to the end.
```suggestion
            $"DirectMethod Report for [{this.SenderSource}] and [{this.ReceiverSource.OrDefault()}] ({this.ResultType})" : $"DirectMethod Report for [{this.SenderSource}] ({this.ResultType})";
```

File: test/modules/Modules.Test/TestResultCoordinator/TestReportUtilTest.cs [hunk 23]
```diff
@@ -261,6 +265,48 @@ public void ParseReportMetadataList_ThrowExceptionWhenParseInvalidData()
             Assert.True(exceptionThrown);
         }
 
+        [Fact]
+        public void DirectMethodReportMetadataJsonDeserializeWithReceiverSourceTest()
+        {
+            const string SerializedTestJsonWithReceiverSource = "{ \"TestReportType\": \"DirectMethodReport\", \"SenderSource\": \"directMethodSender1.send\"," +
```
Comment (philipktlin) [inline test/modules/Modules.Test/TestResultCoordinator/TestReportUtilTest.cs]: please use @ for json string literal to read easily.

File: test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGeneratorTest.cs [hunk 24]
```diff
@@ -17,210 +18,40 @@ namespace Modules.Test.TestResultCoordinator.Reports.DirectMethod
...
-                        new DateTime(2020, 1, 1, 9, 10, 24, 15),
-                        new DateTime(2020, 1, 1, 9, 10, 24, 17),
-                        new DateTime(2020, 1, 1, 9, 10, 25, 20) // NetworkOffFailure
-                        // Mismatch Success is the missing 9
-                        // MismatchFailure is the presence of 11 in the actualStoreValues
-                    },
-                    10, 3, 2, 1, 1, 1, 1, 1, 1, false
-                }
-            };
+        public static IEnumerable<object[]> GetCreateReportDataForTwoSources => DirectMethodReportDataWithSenderAndReceiverSource.GetCreateReportData;
```
Comment (philipktlin) [inline test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGeneratorTest.cs]: I think you can directly use DirectMethodReportDataWithSenderAndReceiverSource.GetCreateReportData in MemberData attribute.

File: test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGeneratorTest.cs [hunk 25]
```diff
@@ -436,37 +240,92 @@ public async Task TestCreateReportAsyncWithEmptyResults()
...
+            Assert.Equal(expectedNetworkOnFailure, report.NetworkOnFailure);
+            Assert.Equal(expectedNetworkOffFailure, report.NetworkOffFailure);
+            Assert.Equal(expectedMismatchSuccess, report.MismatchSuccess);
+            Assert.Equal(expectedMismatchFailure, report.MismatchFailure);
+            Assert.Equal(expectedIsPassed, report.IsPassed);
+        }
+
+        [Theory]
+        [MemberData(nameof(GetCreateReportDataForOneSource))]
+        public async Task TestCreateReportWithNoReceiverResultsAsync(
```
Comment (philipktlin) [inline test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGeneratorTest.cs]:
```suggestion
        public async Task TestCreateReportWithSenderResultsOnlyAsync(
```

File: test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportDataWithSenderAndReceiverSource.cs [hunk 26]
```diff
@@ -0,0 +1,183 @@
...
+                        new DateTime(2020, 1, 1, 9, 10, 24, 15)
+                    },
+                    10, 6, 1, 0, 0, 0, 0, 0, 0, true
+                },
+                new object[]
+                {
+                    // NetworkOnToleratedSuccess test
+                    Enumerable.Range(1, 7).Select(v => v.ToString()),
+                    new[] { "1", "2", "4", "5", "6", "7" },
+                    new int[] { 200, 200, 500, 200, 200, 200, 200 },
```
Comment (philipktlin) [inline test/modules/Modules.Test/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportDataWithSenderAndReceiverSource.cs]: can you use HttpStatusCode.Ok and HttpStatusCode.InternalServerError; and use new List<HttpStatusCode>.

File: test/modules/Modules.Test/TestResultCoordinator/TestReportUtilTest.cs [hunk 27]
```diff
@@ -261,6 +265,48 @@ public void ParseReportMetadataList_ThrowExceptionWhenParseInvalidData()
             Assert.True(exceptionThrown);
         }
 
+        [Fact]
+        public void DirectMethodReportMetadataJsonDeserializeWithReceiverSourceTest()
```
Comment (philipktlin) [inline test/modules/Modules.Test/TestResultCoordinator/TestReportUtilTest.cs]: why do you test deserialization of direct method report metadata json? you should test ParseReportMetadataList in TestReportUtil.  Please follow other tests in this file, e.g. ParseReportMetadataList_ParseDirectMethodTestReportMetadata; and name the test consistently.

File: test/modules/Modules.Test/TestResultCoordinator/TestReportUtilTest.cs [hunk 28]
```diff
@@ -261,6 +265,48 @@ public void ParseReportMetadataList_ThrowExceptionWhenParseInvalidData()
...
+            Assert.Equal("directMethodSender1.send", directMethodReportMetadata.SenderSource);
+            Assert.Equal(new TimeSpan(0, 0, 0, 0, 5), directMethodReportMetadata.TolerancePeriod);
+            Assert.Equal(TestReportType.DirectMethodReport, directMethodReportMetadata.TestReportType);
+            Assert.True(!directMethodReportMetadata.ReceiverSource.HasValue);
+        }
+
+        [Fact]
+        public void DirectMethodTestResultJsonSerializeTest()
+        {
+            DirectMethodTestResult directMethodTestResult = new DirectMethodTestResult(
```
Comment (philipktlin) [inline test/modules/Modules.Test/TestResultCoordinator/TestReportUtilTest.cs]: you should test serialize and deserialize in 2 tests; and you should not directly test JsonConvert.SerializeObjcet and JsonConvert.DeserializeObject.  You need to test the code in our codebase.

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs [hunk 29]
```diff
@@ -80,22 +80,28 @@ public async Task<ITestResultReport> CreateReportAsync()
...
+                        if (additionalCountsAndHasResults.MismatchFailure > 0 || additionalCountsAndHasResults.MismatchSuccess > 0)
+                        {
+                            continue;
+                        }
                     }
                 }
 
-                AdditionalCountsAndHasResults additionalCountsAndHasResultsSendery =
-                    await this.SenderOnlyLogic(dmSenderTestResult, networkControllerStatus, isWithinTolerancePeriod, this.SenderTestResults);
+                DirectMethodReportGeneratorMetadata additionalCountsAndHasResultsSendery =
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs]: typo
```suggestion
                DirectMethodReportGeneratorMetadata additionalCountsAndHasResultsSender =
```

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs [hunk 30]
```diff
@@ -82,35 +83,33 @@ public async Task<ITestResultReport> CreateReportAsync()
...
+                networkOffFailure += metadata.NetworkOffFailure;
+                hasSenderResult = metadata.HasSenderResult;
             }
 
             while (hasReceiverResult)
             {
-                DirectMethodReportGeneratorMetadata additionalCountsAndHasResultsForMismatchFailure = await this.MismatchFailureCase();
-                mismatchFailure += additionalCountsAndHasResultsForMismatchFailure.MismatchFailure;
-                hasReceiverResult = additionalCountsAndHasResultsForMismatchFailure.HasReceiverResult;
+                metadata = await this.MismatchFailureCase();
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs]: nit: rename this method; Mismatch is adjective, right? normally starts with a verb.

File: test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs [hunk 31]
```diff
@@ -71,6 +71,7 @@ public async Task<ITestResultReport> CreateReportAsync()
 
             bool hasSenderResult = await this.SenderTestResults.MoveNextAsync();
             bool hasReceiverResult = await this.ReceiverTestResults.Match(async x => await x.MoveNextAsync(), () => Task.FromResult(false));
+            DirectMethodReportGeneratorMetadata metadata;
```
Comment (philipktlin) [inline test/modules/TestResultCoordinator/Reports/DirectMethod/DirectMethodReportGenerator.cs]: rename reportGeneratorMetadata

File: test/modules/TwinTester/Program.cs [hunk 32]
```diff
@@ -20,6 +20,9 @@ static async Task Main()
 
             Logger.LogInformation($"Starting twin tester with the following settings:\r\n{Settings.Current}");
 
+            Logger.LogInformation($"Load gen delay start for {Settings.Current.TestStartDelay}.");
```
Comment (philipktlin) [inline test/modules/TwinTester/Program.cs]: what is this change

Comment (dylanbronson) [reply test/modules/TwinTester/Program.cs]: whoops. Removed.
